### ISSUE TITLE
migrate2rocky.sh: Failed update cannot be continued

### ISSUE THREAD

**Issue by lbdroid** [rocky-linux/rocky-tools#69]
I ran into an issue where a bad repository resulted in the following errors;

- Errors during downloading metadata for repository
- Can't enable modules
- An error occurred while we were attempting to convert your system to Rocky Linux. Your system may be unstable. Script will now exit to prevent possible damage.

After correcting the error (removing the bad repository) and attempting to rerun the script to continue the migration, received the following errors;

- Found a full or partial RockyLinux install already in place.  Aborting
- An error occurred while we were attempting to convert your system to Rocky Linux. Your system may be unstable. Script will now exit to prevent possible damage.

So the system was left in a broken state with no tools available to either complete the migration or roll back.

**Comment by pajamian** [comment]
Can you provide the migrate2rocky.log file from the first failed migration, please?

**Comment by pajamian** [comment]
Was this a migration from OracleLinux per chance?

**Comment by lbdroid** [comment]
It was from CentOS 8.4 fully up to date as of about 3 minutes before running the migration.

```
Preparing to migrate CentOS Linux 8 to Rocky Linux 8.

Determining repository names for CentOS Linux 8.....

Found the following repositories which map from CentOS Linux 8 to Rocky Linux 8:
CentOS Linux 8  Rocky Linux 8
appstream       appstream
baseos          baseos
extras          extras

Getting system package names for CentOS Linux 8..........

Found the following system packages which map from CentOS Linux 8 to Rocky Linux 8:
CentOS Linux 8        Rocky Linux 8
centos-logos-ipa      rocky-logos-ipa
centos-backgrounds    rocky-backgrounds
centos-gpg-keys       rocky-gpg-keys
centos-logos          rocky-logos
centos-indexhtml      rocky-indexhtml
centos-linux-release  rocky-release
centos-logos-httpd    rocky-logos-httpd
centos-linux-repos    rocky-repos

Getting list of installed system packages.

We will replace the following CentOS Linux 8 packages with their Rocky Linux 8 equivalents
Packages to be Removed  Packages to be Installed
centos-backgrounds      rocky-backgrounds
centos-gpg-keys         rocky-gpg-keys
centos-logos            rocky-logos
centos-indexhtml        rocky-indexhtml
centos-linux-release    rocky-release
centos-logos-httpd      rocky-logos-httpd
centos-linux-repos      rocky-repos

In addition to the above the following system packages will be removed:
centos-linux-release
centos-linux-release

Getting a list of enabled modules for the system repositories.

Excluding modules:
libselinux-python:2.8

Found the following modules to re-enable at completion:
container-tools:rhel8
httpd:2.4
javapackages-runtime:201801
llvm-toolset:rhel8
mariadb:10.3
nginx:1.14
nodejs:10
perl:5.26
perl-DBD-MySQL:4.046
perl-DBD-SQLite:1.58
perl-DBI:1.641
perl-IO-Socket-SSL:2.066
perl-libwww-perl:6.34
php:7.4
python27:2.7
python36:3.6
satellite-5-client:1.0
virt:rhel

Running dnf update before we attempt the migration.
Last metadata expiration check: 1:57:37 ago on Mon 05 Jul 2021 01:44:02 PM EDT.
Dependencies resolved.
Nothing to do.
Complete!
Added rockyappstream repo from https://dl.rockylinux.org/pub/rocky/8/AppStream/x86_64/os/
Added rockybaseos repo from https://dl.rockylinux.org/pub/rocky/8/BaseOS/x86_64/os/
rockyappstream                                  5.4 MB/s | 7.8 MB     00:01    
rockybaseos                                     4.0 MB/s | 3.5 MB     00:00    
Last metadata expiration check: 0:00:01 ago on Mon 05 Jul 2021 03:41:46 PM EDT.
> > > ================================================================================
 Package                 Arch      Version              Repository         Size
================================================================================
Installing:
 rocky-backgrounds       noarch    84.5-8.el8           rockyappstream    330 k
 rocky-gpg-keys          noarch    8.4-29.el8           rockybaseos        11 k
 rocky-indexhtml         noarch    8.0-0.el8            rockybaseos       266 k
 rocky-logos             x86_64    84.5-8.el8           rockybaseos       328 k
 rocky-logos-httpd       noarch    84.5-8.el8           rockybaseos        22 k
 rocky-release           noarch    8.4-29.el8           rockybaseos        20 k
 rocky-repos             noarch    8.4-29.el8           rockybaseos        13 k
Removing:
 centos-backgrounds      noarch    85.8-1.el8           @appstream         23 M
 centos-gpg-keys         noarch    1:8-2.el8            @BaseOS           3.3 k
 centos-indexhtml        noarch    8.0-0.el8            @AppStream        505 k
 centos-linux-release    noarch    8.4-1.2105.el8       @baseos            25 k
 centos-linux-repos      noarch    8-2.el8              @BaseOS            26 k
 centos-logos            x86_64    85.8-1.el8           @baseos           716 k
 centos-logos-httpd      noarch    85.8-1.el8           @baseos           197 k

Transaction Summary
================================================================================
Install  7 Packages
Remove   7 Packages

Total download size: 989 k
Downloading Packages:
(1/7): rocky-gpg-keys-8.4-29.el8.noarch.rpm     155 kB/s |  11 kB     00:00    
(2/7): rocky-backgrounds-84.5-8.el8.noarch.rpm  2.1 MB/s | 330 kB     00:00    
(3/7): rocky-indexhtml-8.0-0.el8.noarch.rpm     1.4 MB/s | 266 kB     00:00    
(4/7): rocky-logos-httpd-84.5-8.el8.noarch.rpm  685 kB/s |  22 kB     00:00    
(5/7): rocky-repos-8.4-29.el8.noarch.rpm        710 kB/s |  13 kB     00:00    
(6/7): rocky-release-8.4-29.el8.noarch.rpm      863 kB/s |  20 kB     00:00    
(7/7): rocky-logos-84.5-8.el8.x86_64.rpm        2.0 MB/s | 328 kB     00:00    
--------------------------------------------------------------------------------
Total                                           4.1 MB/s | 989 kB     00:00     
warning: /var/cache/dnf/rockyappstream-a14182fcf29cfb50/packages/rocky-backgrounds-84.5-8.el8.noarch.rpm: Header V4 RSA/SHA256 Signature, key ID 6d745a60: NOKEY
rockyappstream                                  1.6 MB/s | 1.6 kB     00:00    
Importing GPG key 0x6D745A60:
 Userid     : "Release Engineering <infrastructure@rockylinux.org>"
 Fingerprint: 7051 C470 A929 F454 CEBE 37B7 15AF 5DAC 6D74 5A60
 From       : /tmp/tmp.Axxm1e7jjl/gpg/RPM-GPG-KEY-rockyofficial
Key imported successfully
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                        1/1 
  Running scriptlet: rocky-logos-84.5-8.el8.x86_64                          1/1 
  Installing       : rocky-logos-84.5-8.el8.x86_64                         1/14 
  Running scriptlet: rocky-logos-84.5-8.el8.x86_64                         1/14 
  Installing       : rocky-gpg-keys-8.4-29.el8.noarch                      2/14 
  Installing       : rocky-release-8.4-29.el8.noarch                       3/14 
  Installing       : rocky-repos-8.4-29.el8.noarch                         4/14 
  Installing       : rocky-backgrounds-84.5-8.el8.noarch                   5/14 
  Installing       : rocky-logos-httpd-84.5-8.el8.noarch                   6/14 
  Installing       : rocky-indexhtml-8.0-0.el8.noarch                      7/14 
  Erasing          : centos-linux-release-8.4-1.2105.el8.noarch            8/14 
  Erasing          : centos-linux-repos-8-2.el8.noarch                     9/14 
  Erasing          : centos-backgrounds-85.8-1.el8.noarch                 10/14 
  Erasing          : centos-logos-85.8-1.el8.x86_64                       11/14 
  Running scriptlet: centos-logos-85.8-1.el8.x86_64                       11/14 
  Erasing          : centos-gpg-keys-1:8-2.el8.noarch                     12/14 
  Erasing          : centos-logos-httpd-85.8-1.el8.noarch                 13/14 
  Erasing          : centos-indexhtml-8.0-0.el8.noarch                    14/14 
  Running scriptlet: rocky-logos-84.5-8.el8.x86_64                        14/14 
  Running scriptlet: centos-indexhtml-8.0-0.el8.noarch                    14/14 
  Verifying        : rocky-backgrounds-84.5-8.el8.noarch                   1/14 
  Verifying        : rocky-gpg-keys-8.4-29.el8.noarch                      2/14 
  Verifying        : rocky-indexhtml-8.0-0.el8.noarch                      3/14 
  Verifying        : rocky-logos-84.5-8.el8.x86_64                         4/14 
  Verifying        : rocky-logos-httpd-84.5-8.el8.noarch                   5/14 
  Verifying        : rocky-release-8.4-29.el8.noarch                       6/14 
  Verifying        : rocky-repos-8.4-29.el8.noarch                         7/14 
  Verifying        : centos-backgrounds-85.8-1.el8.noarch                  8/14 
  Verifying        : centos-gpg-keys-1:8-2.el8.noarch                      9/14 
  Verifying        : centos-indexhtml-8.0-0.el8.noarch                    10/14 
  Verifying        : centos-linux-release-8.4-1.2105.el8.noarch           11/14 
  Verifying        : centos-linux-repos-8-2.el8.noarch                    12/14 
  Verifying        : centos-logos-85.8-1.el8.x86_64                       13/14 
  Verifying        : centos-logos-httpd-85.8-1.el8.noarch                 14/14 
Installed products updated.

Installed:
  rocky-backgrounds-84.5-8.el8.noarch      rocky-gpg-keys-8.4-29.el8.noarch     
  rocky-indexhtml-8.0-0.el8.noarch         rocky-logos-84.5-8.el8.x86_64        
  rocky-logos-httpd-84.5-8.el8.noarch      rocky-release-8.4-29.el8.noarch      
  rocky-repos-8.4-29.el8.noarch           
Removed:
  centos-backgrounds-85.8-1.el8.noarch                                          
  centos-gpg-keys-1:8-2.el8.noarch                                              
  centos-indexhtml-8.0-0.el8.noarch                                             
  centos-linux-release-8.4-1.2105.el8.noarch                                    
  centos-linux-repos-8-2.el8.noarch                                             
  centos-logos-85.8-1.el8.x86_64                                                
  centos-logos-httpd-85.8-1.el8.noarch                                          

Complete!
Last metadata expiration check: 0:00:07 ago on Mon 05 Jul 2021 03:41:46 PM EDT.
> Leaving Shell

Removing dnf cache
Ensuring repos are enabled before the package swap
Enabling modules

Rocky Linux 8 - AppStream                       4.6 MB/s | 7.8 MB     00:01    
Rocky Linux 8 - BaseOS                          4.9 MB/s | 3.5 MB     00:00    
Rocky Linux 8 - Extras                           12 kB/s | 3.8 kB     00:00    
bintray--ookla-rhel                              17  B/s |  10  B     00:00    
Errors during downloading metadata for repository 'bintray--ookla-rhel':
  - Status code: 403 for https://ookla.bintray.com/rhel/repodata/repomd.xml (IP: 52.88.131.165)
Error: Failed to download metadata for repo 'bintray--ookla-rhel': Cannot download repomd.xml: Cannot download repodata/repomd.xml: All mirrors were tried

Can't enable modules container-tools:rhel8 httpd:2.4 javapackages-runtime:201801 llvm-toolset:rhel8 mariadb:10.3 nginx:1.14 nodejs:10 perl:5.26 perl-DBD-MySQL:4.046 perl-DBD-SQLite:1.58 perl-DBI:1.641 perl-IO-Socket-SSL:2.066 perl-libwww-perl:6.34 php:7.4 python27:2.7 python36:3.6 satellite-5-client:1.0 virt:rhel

An error occurred while we were attempting to convert your system to Rocky Linux. Your system may be unstable. Script will now exit to prevent possible damage.

```

**Comment by pajamian** [comment]
Manually running the remaining steps is the best way to go about completing a failed migration, I'm glad you got your system migrated that way.

So the actual error here is that it can't download the repository metadata for your 3rd-party module.  I would have thought that would have been caught in the initial dnf update and failed then  before changes were made, but it looks like dnf had the metadata cached at that time, and the script, at a later stage, clears out the dnf cache, thereby forcing it to try to downlod the metadata again in the module enable stage.

I think the best solution here is to dump the cache before the update, not after, then the update should catch issues like this before any changes are made that would result in a partial migration.

**Comment by lbdroid** [comment]
Well yes, it would help if the error was caught before the migration started for sure, however I think that really the bigger issue is that the migration can't be resumed once the cause of the error is found and corrected.

It wouldn't need a complicated mechanism, the migration already follows a sequence of phases, so when it fails, output something to the effect of "Migration failed during phase X. After correcting above listed errors, please resume migration by adding parameter --phase=X", and the parameter would cause the migration to (a) recognize that there was a partial migration [not fail due to Rocky->Rocky migration], and (b) have the script skip over changes during phases < X.

**Comment by pajamian** [comment]
The issue with that is that there is infomation that needs to be collected prior to the migration that will not necessarily be reliable after a failed migration.  If the script is run again it will not have all that information and it is impossible to resume the migration at any particular stage due to this.

What is planned, though, is to have a sort of "attempt to fix a failed migration" option, that will basically attempt to force install a few essential system packages, run the distro-sync and then the EFI fix (if appropriate).  It will not necessarily restore an identical system to that which was there to begin with, but it should be able to restore the system to a running, bootable state, and further alterations can then be made by the user to fix any remaining anomalies.

It is still important to try to detect things such as this that might cause issues and stop the migration before it starts.

**Comment by lbdroid** [comment]
Hmm... if the values can't be reliably collected after a partial migration, then it would be easy to drop them into a resume script that's written along with the log file.

**Comment by pajamian** [comment]
This could work, but I'm tending to prefer a more generic brute-force approach, simply because I think it would have a greater chance of succeeding where the previous attempt failed.  This is about trying to get a system that has partially migrated to a state where it is booting and running again, which has a lower bar than "try to get a system that is in exactly the same state as previously."

**Comment by pajamian** [comment]
Fixed with PR #76 
Will continue conversation about failed migration recovery in issue #57
